addons/account/wizard/account_payment_register.py

```diff
@@ -116,9 +116,54 @@ class AccountPaymentRegister(models.TransientModel):
         help="Technical field used to know whether the field `partner_bank_id` needs to be required or not in the payments form views")
     country_code = fields.Char(related='company_id.account_fiscal_country_id.code', readonly=True)
 
+    # == Early Payment Discount ==
+    has_epd_and_is_in_time = fields.Boolean(
+        compute='_compute_has_epd_and_is_in_time',
+        readonly=True,
+        store=False,
+        help="Indicates whether or not an early payment discount is available and if the conditions required to activate it are fulfilled",
+    )
+
+    show_epd_button = fields.Boolean(compute="_compute_show_epd_button")
+
     # -------------------------------------------------------------------------
     # HELPERS
     # -------------------------------------------------------------------------
+    @api.onchange('amount')
+    def _on_change_amount(self):
+        for wizard in self:
+            move_id = wizard.line_ids.move_id
+            if wizard.has_epd_and_is_in_time and wizard.amount == move_id.invoice_early_pay_amount_after_discount:
+                wizard.writeoff_label = _('Early Payment Discount')
+                invoice_payment_term = wizard.line_ids.move_id.invoice_payment_term_id
+                wizard.writeoff_account_id = invoice_payment_term.get_early_payment_discount_account(move_id.is_in_invoice)
+                wizard.payment_difference_handling = 'reconcile'
+
+    @api.onchange('payment_date')
+    def _on_change_payment_date(self):
+        for wizard in self:
+            wizard.has_epd_and_is_in_time = wizard.line_ids.move_id.is_eligible_for_early_discount(wizard.payment_date)
+            if wizard.has_epd_and_is_in_time and wizard.line_ids.move_id.is_in_invoice:
+                wizard.amount = wizard.line_ids.move_id.invoice_early_pay_amount_after_discount
```

Comment:
I'm afraid we want to avoid onchanges anymore... Can you make them computed fields instead?